src/ClientBulkWrite.php

```diff
@@ -0,0 +1,241 @@
+        $update = $this->builderEncoder->encodeIfSupported($update);
+
+        if (! is_first_key_operator($update) && ! is_pipeline($update)) {
+            throw new InvalidArgumentException('Expected update operator(s) or non-empty pipeline for $update');
+        }
+
+        if (isset($options['arrayFilters']) && ! is_array($options['arrayFilters'])) {
+            throw InvalidArgumentException::invalidType('"arrayFilters" option', $options['arrayFilters'], 'array');
+        }
+
+        if (isset($options['collation']) && ! is_document($options['collation'])) {
+            throw InvalidArgumentException::expectedDocumentType('"collation" option', $options['collation']);
+        }
+
+        if (isset($options['hint']) && ! is_string($options['hint']) && ! is_document($options['hint'])) {
+            throw InvalidArgumentException::expectedDocumentOrStringType('"hint" option', $options['hint']);
+        }
+
+        if (isset($options['upsert']) && ! is_bool($options['upsert'])) {
+            throw InvalidArgumentException::invalidType('"upsert" option', $options['upsert'], 'boolean');
+        }
+
+        $this->bulkWriteCommand->updateMany($this->namespace, $filter, $update, $options);
+
+        return $this;
+    }
+
+    public function updateOne(array|object $filter, array|object $update, array $options = []): self
+    {
+        $filter = $this->builderEncoder->encodeIfSupported($filter);
+        $update = $this->builderEncoder->encodeIfSupported($update);
+
+        if (! is_first_key_operator($update) && ! is_pipeline($update)) {
+            throw new InvalidArgumentException('Expected update operator(s) or non-empty pipeline for $update');
+        }
+
+        if (isset($options['arrayFilters']) && ! is_array($options['arrayFilters'])) {
+            throw InvalidArgumentException::invalidType('"arrayFilters" option', $options['arrayFilters'], 'array');
+        }
+
+        if (isset($options['collation']) && ! is_document($options['collation'])) {
+            throw InvalidArgumentException::expectedDocumentType('"collation" option', $options['collation']);
+        }
+
+        if (isset($options['hint']) && ! is_string($options['hint']) && ! is_document($options['hint'])) {
+            throw InvalidArgumentException::expectedDocumentOrStringType('"hint" option', $options['hint']);
+        }
+
+        if (isset($options['sort']) && ! is_document($options['sort'])) {
+            throw InvalidArgumentException::expectedDocumentType('"sort" option', $options['sort']);
+        }
+
+        if (isset($options['upsert']) && ! is_bool($options['upsert'])) {
+            throw InvalidArgumentException::invalidType('"upsert" option', $options['upsert'], 'boolean');
+        }
+
+        $this->bulkWriteCommand->updateOne($this->namespace, $filter, $update, $options);
+
+        return $this;
+    }
+
+    public function withCollection(Collection $collection): self
```

Comment:
You can add the upcoming `#[NoDiscard]` attribute on this method, as it returns a new instance. @alcaeus approved the [RFC](https://wiki.php.net/rfc/marking_return_value_as_important), so he would agree.

This would prevent people from thinking this is mutating the object.

Reply:
Should that also be done for `createWithCollection()`, since it also returns a new instance? `#[NoDiscard]` seems like it'd be appropriate for _any_ factory method.

Reply:
I assume this will also require baselining the UndefinedAttributeClass error in Psalm.

Reply:
Since adding `#[NoDiscard]` later could be seen as a BC break, I think it's better to add the attribute now and silence the Psalm errors. Unless people are actively trying to read this particular attribute, it will have no effect until PHP 8.5 is released. Even then, the polyfill will allow libraries (such as Psalm) to look for the attribute on earlier PHP versions and provide the same functionality.

Reply:
Attribute added.